listenbrainz/spark/spark_reader.py

```diff
@@ -0,0 +1,85 @@
+import json
+import logging
+import pika
+import time
+import ujson
+
+from flask import current_app
+from listenbrainz import utils
+from listenbrainz.db import user as db_user, stats as db_stats
+from listenbrainz.webserver import create_app
+from listenbrainz.db.exceptions import DatabaseException
+from listenbrainz import config
+import sqlalchemy
+
+class SparkReader:
+    def __init__(self):
+        self.app = create_app() # creating a flask app for config values and logging to Sentry
+
+
+    def init_rabbitmq_connection(self):
+        """ Initializes the connection to RabbitMQ.
+
+        Note: this is a blocking function which keeps retrying if it fails
+        to connect to RabbitMQ
+        """
+        self.connection = utils.connect_to_rabbitmq(
+            username=current_app.config['RABBITMQ_USERNAME'],
+            password=current_app.config['RABBITMQ_PASSWORD'],
+            host=current_app.config['RABBITMQ_HOST'],
+            port=current_app.config['RABBITMQ_PORT'],
+            virtual_host=current_app.config['RABBITMQ_VHOST'],
+            error_logger=current_app.logger.error,
+        )
+
+
+    def callback(self, ch, method, properties, body):
+        """ Handle the data received from the queue and works accordingly.
+        """
+        data = ujson.loads(body)
+        user_name = next(iter(data[0]))
```

Comment:
What exactly does this mean? I wonder if we can make it a bit clearer?

Reply:
This is used to retrieve the first key of a dictionary, which would always be the user name in this case.

Reply:
I think this should just be data.keys()[0]?

Reply:
In anything other than python 3.7, dictionaries are not guaranteed to be in any specific order. Especially when reading from json I would be very careful about this. What does the format of the body look like? Is it always {username: {somedata}, otherusername: {moredata}} ? How many user datas are returned? always 1 or more?
You could use `for username, stats in data.items()`, which would work even if there was only data for 1 username. It would also clean up the accesses later in this method